Issue #1488 - added build duration and tweaked build counting in describer

soltysh opened the pull request:
@bparees PTAL, I've added duration and tweaked the build counting :wink: and now we can safely close #1488

soltysh commented:
This is how it looks with this:
```
[vagrant@openshiftdev origin]$ osc describe buildconfig ruby-sample-build
Name:                           ruby-sample-build
Created:                        2015-04-07 11:24:07 +0000 UTC
Labels:                         name=ruby-sample-build,template=application-template-stibuild
Latest Version:                 12
Strategy:                       STI
Builder Image:                  openshift/ruby-20-centos7
Incremental Build:              yes
Source Type:                    Git
URL:                            git://github.com/openshift/ruby-hello-world.git
Output to:                      origin-ruby-sample:latest
Output Spec:                    <none>
Webhook Generic:                https://10.0.2.15:8443/osapi/v1beta1/buildConfigHooks/ruby-sample-build/secret101/generic
Webhook Github:                 https://10.0.2.15:8443/osapi/v1beta1/buildConfigHooks/ruby-sample-build/secret101/github
Image Repository Trigger:       ruby-20-centos7
- Tag:                          latest
- Image:                        openshift/ruby-20-centos7
- LastTriggeredImageID:         openshift/ruby-20-centos7:latest
Builds(Name,Status,Duration,Creation Time)
- ruby-sample-build-12 Failed (31s) 2015-04-07 11:25:01 +0000 UTC
- ruby-sample-build-11 Failed (21s) 2015-04-07 11:24:32 +0000 UTC
- ruby-sample-build-10 Failed (2m6s) 2015-04-07 11:24:31 +0000 UTC
- ruby-sample-build-9 Failed (2m6s) 2015-04-07 11:24:30 +0000 UTC
- ruby-sample-build-8 Failed (2m17s) 2015-04-07 11:24:29 +0000 UTC
- ruby-sample-build-7 Failed (21s) 2015-04-07 11:24:28 +0000 UTC
- ruby-sample-build-6 Failed (2m17s) 2015-04-07 11:24:27 +0000 UTC
- ruby-sample-build-5 Failed (2m17s) 2015-04-07 11:24:25 +0000 UTC
- ruby-sample-build-4 Failed (1m58s) 2015-04-07 11:24:23 +0000 UTC
- ruby-sample-build-3 Failed (21s) 2015-04-07 11:24:22 +0000 UTC
```

bparees commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -283,17 +283,16 @@ func (d *BuildConfigDescriber) Describe(namespace, name string) (string, error)
 			build := sortedBuilds[len(sortedBuilds)-1-i]
-			fmt.Fprintf(out, fmt.Sprintf("- %s %s %v\n", build.Name, build.Status, build.CreationTimestamp.Rfc3339Copy().Time))
-			c++
+			duration := build.CompletionTimestamp.Rfc3339Copy().Time.Sub(build.StartTimestamp.Rfc3339Copy().Time)
```
StartTimestamp could be nil.  For that matter so could CompletionTimestamp so i'm surprised this is working?

In anycase you need to look at this logic to see how complex this really needs to be do to it right (which is why I didn't do it :) )

https://github.com/openshift/origin/blob/master/pkg/cmd/cli/describe/describer.go#L114

soltysh replied:
If so, I'll extract a method for doing that, thx.

smarterclayton replied:
See pkg/api/graph/deployment.go, if you extract extract it from both places.

----- Original Message -----
> >  		for i := range sortedBuilds {
> >  			// iterate backwards so we're printing the newest items first
> >  			build := sortedBuilds[len(sortedBuilds)-1-i]
> > -			fmt.Fprintf(out, fmt.Sprintf("- %s %s %v\n", build.Name, build.Status,
> > build.CreationTimestamp.Rfc3339Copy().Time))
> > -			c++
> > +			duration :=
> > build.CompletionTimestamp.Rfc3339Copy().Time.Sub(build.StartTimestamp.Rfc3339Copy().Time)
> 
> If so, I'll extract a method for doing that, thx.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/1624/files#r27883923

smarterclayton replied:
Also see https://github.com/openshift/origin/pull/1621

----- Original Message -----
> See pkg/api/graph/deployment.go, if you extract extract it from both places.
> 
> ----- Original Message -----
> > >  		for i := range sortedBuilds {
> > >  			// iterate backwards so we're printing the newest items first
> > >  			build := sortedBuilds[len(sortedBuilds)-1-i]
> > > -			fmt.Fprintf(out, fmt.Sprintf("- %s %s %v\n", build.Name,
> > > build.Status,
> > > build.CreationTimestamp.Rfc3339Copy().Time))
> > > -			c++
> > > +			duration :=
> > > build.CompletionTimestamp.Rfc3339Copy().Time.Sub(build.StartTimestamp.Rfc3339Copy().Time)
> > 
> > If so, I'll extract a method for doing that, thx.
> > 
> > ---
> > Reply to this email directly or view it on GitHub:
> > https://github.com/openshift/origin/pull/1624/files#r27883923
>

smarterclayton replied:
That section is insane, and means the server should be returning duration so that clients don't have to implement it. :)

soltysh replied:
Yeah, my idea would be to either:
1. add `Duration` field to `Build` object that will be filled upon get (I don't think we should store that)
2. replace `CompletionTimestamp` with `Duration`, IMHO these two make more sense than having start and end time.
Personally I'm leaning towards the latter, what do you think @bparees @smarterclayton ?

bparees replied:
i'd prefer #1.  Replacing completionTimestamp doesn't make sense as we are calculating various "durations" depending on the state of the build (duration waiting for a pod, duration actually running) and we want a duration value for running builds, not just completed ones.

Also be aware that the web console uses all these timestamp values and does similar logic, so if you change the behavior, the webconsole should be updated as well.

soltysh replied:
Yeah, I was about to write that, after carefully studying the code you've pointed to, where I see you also print waiting time as well.

soltysh replied:
The Duration here will specifically be of the build itself, meaning if it hasn't started yet (CcreationTimestamp==0) it's Duration will be 0. In all other cases it will always show the Duration for the moment of GET, which obviously mean it'll change for running builds, but will be constant for Finished, Cancelled etc.

bparees replied:
that would be a change for waiting builds since today we have the duration for builds that haven't started yet...what's your plan for that?

soltysh replied:
That waiting information is only in build describe and I'll leave it there as is, since I'm not planning to add descriptive field about the duration it'll contain just the build duration. So in a case of waiting duration will be 0.

soltysh replied:
Something like that:
```
Builds:
  Name                  Status          Duration        Creation Time
  ruby-sample-build-23  Complete        50s             2015-04-08 13:26:53 +0000 UTC
  ruby-sample-build-22  Pending         0               2015-04-08 13:18:52 +0000 UTC
  ruby-sample-build-20  Failed          31s             2015-04-08 13:18:51 +0000 UTC
```

smarterclayton commented:
Use tab writer to align the columns - the uneven output is ugly.

smarterclayton commented:
Lowercase the status, and use formatRelativeTime() instead of displaying the exact time.

smarterclayton commented:
Drop the dash in front of each line and do an indent consistent with the rest of the page (usually \t but smaller is ok)

soltysh commented:
I've addressed all the comments except for the `formatRelativeTime()` I'm waiting for #1621. Current output is following:
```
Builds:
  Name                  Status          Duration                Creation Time
  ruby-sample-build-26  pending         waiting for 2s          2015-04-08 14:42:33 +0000 UTC
  ruby-sample-build-25  running         running for 16s         2015-04-08 14:42:07 +0000 UTC
  ruby-sample-build-24  complete        50s                     2015-04-08 14:39:27 +0000 UTC
  ruby-sample-build-23  complete        50s                     2015-04-08 13:26:53 +0000 UTC
  ruby-sample-build-22  failed          1m0s                    2015-04-08 13:18:52 +0000 UTC
```

bparees commented on pkg/cmd/cli/describe/describer.go:
```diff
@@ -111,30 +111,25 @@ func (d *BuildDescriber) Describe(namespace, name string) (string, error) {
+	t := util.Now().Rfc3339Copy()
+	if build.StartTimestamp == nil {
+		// time a new build has been waiting for its pod to be created so it can run
+		return fmt.Sprintf("waiting for %v", t.Sub(build.CreationTimestamp.Rfc3339Copy().Time))
```
you're going to output "waiting for XXXs" for builds that were canceled before starting.  Should output "waited"

soltysh replied:
OK, I'll prepare test cases for that and then I'll change it accordingly.

bparees commented:
one nit, otherwise this looks much nicer, thanks!

soltysh commented:
@bparees addressed that one comment and added tests for that, squashed it all, ready to merge.

mfojtik commented:
LGTM

bparees commented:
latm [merge]

soltysh commented:
Rebased.

soltysh commented:
re-[test] I don't understand why server died when testing UI.

smarterclayton commented:
[merge]
